File: src/main/kotlin/com/kursx/parser/fb2/Binary.java
package com.kursx.parser.fb2;

import com.sun.xml.bind.v2.runtime.unmarshaller.Base64Data;
import org.jetbrains.annotations.NotNull;

import org.w3c.dom.NamedNodeMap;
import org.w3c.dom.Node;

//http://www.fictionbook.org/index.php/Элемент_binary
@SuppressWarnings("unused")
public class Binary extends IdElement {

    protected String contentType;
    protected String binary;

    public Binary() {
    }

    Binary(Node node) {
        super(node);
        binary = node.getTextContent();
        NamedNodeMap map = node.getAttributes();
        for (int index = 0; index < map.getLength(); index++) {
            Node attr = map.item(index);
            if ("content-type".equals(attr.getNodeName())) {
                contentType = attr.getNodeValue();
            }
        }
    }

    public String getContentType() {
        return contentType;
    }

    @NotNull
    public String getBinary() {
        return binary;
    }

    @NotNull
    @Override
    public String getId() {
        return super.getId();
    }

    public void setContentType(String contentType) {
        this.contentType = contentType;
    }

    public void setBinary(String binary) {
        this.binary = binary;
    }

    @Override
    public String toString() {
        return "Binary{" +
                "contentType='" + contentType + '\'' +
                ", binary='" + binary + '\'' +
                ", id='" + id + '\'' +
                '}';
    }
}

File: src/main/kotlin/com/kursx/parser/fb2/Binary.kt
package com.kursx.parser.fb2

import kotlinx.serialization.Serializable
import kotlinx.serialization.protobuf.ProtoNumber
import org.w3c.dom.Node

//http://www.fictionbook.org/index.php/Элемент_binary
@Suppress("unused")
@Serializable
class Binary : IdElement {
    @ProtoNumber(2)
    var contentType: String? = null
    @ProtoNumber(3)
    lateinit var binary: ByteArray

    constructor()
    internal constructor(node: Node) : super(node) {
        binary = org.apache.commons.codec.binary.Base64().decode(node.textContent)
        val map = node.attributes
        for (index in 0 until map.length) {
            val attr = map.item(index)
            if ("content-type" == attr.nodeName) {
                contentType = attr.nodeValue
            }
        }
    }

    override fun toString(): String {
        return "Binary(contentType=$contentType, binary=${binary.contentToString()})"
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        if (!super.equals(other)) return false

        other as Binary

        if (contentType != other.contentType) return false
        if (!binary.contentEquals(other.binary)) return false

        return true
    }

    override fun hashCode(): Int {
        var result = super.hashCode()
        result = 31 * result + (contentType?.hashCode() ?: 0)
        result = 31 * result + binary.contentHashCode()
        return result
    }


}
